x/precisebank/keeper/send.go

```diff
@@ -19,23 +32,293 @@ func (k Keeper) SendCoins(
+	amt sdkmath.Int,
+) error {
+	// Sufficient balance check is done by bankkeeper.SendCoins(), for both
+	// integer and fractional-only sends. E.g. If fractional balance is
+	// insufficient, it will still incur a integer borrow which will fail if the
+	// sender does not have sufficient integer balance.
+
+	// Load required state: Account old balances
+	senderFracBal := k.GetFractionalBalance(ctx, from)
+	recipientFracBal := k.GetFractionalBalance(ctx, to)
+
+	// -------------------------------------------------------------------------
+	// Pure stateless calculations
+	integerAmt := amt.Quo(types.ConversionFactor())
+	fractionalAmt := amt.Mod(types.ConversionFactor())
+
+	// Account new fractional balances
+	senderNewFracBal, senderNeedsBorrow := subFromFractionalBalance(senderFracBal, fractionalAmt)
+	recipientNewFracBal, recipientNeedsCarry := addToFractionalBalance(recipientFracBal, fractionalAmt)
+
+	// Case #1: Sender borrow, recipient carry
+	if senderNeedsBorrow && recipientNeedsCarry {
+		// Can directly transfer borrow/carry - increase the direct transfer by 1
+		integerAmt = integerAmt.AddRaw(1)
+	}
+
+	// -------------------------------------------------------------------------
+	// Stateful operations for transfers
+
+	// This includes ALL transfers of >= conversionFactor AND Case #1
+	// Full integer amount transfer, including direct transfer of borrow/carry
+	// if any.
+	if integerAmt.IsPositive() {
+		transferCoin := sdk.NewCoin(types.IntegerCoinDenom, integerAmt)
+		if err := k.bk.SendCoins(ctx, from, to, sdk.NewCoins(transferCoin)); err != nil {
+			return k.updateInsufficientFundsError(ctx, from, amt, err)
+		}
+	}
+
+	// Case #2: Sender borrow, NO recipient carry
+	// Sender borrows by transferring 1 integer amount to reserve to account for
+	// lack of fractional balance.
+	if senderNeedsBorrow && !recipientNeedsCarry {
+		borrowCoin := sdk.NewCoin(types.IntegerCoinDenom, sdk.NewInt(1))
+		if err := k.bk.SendCoinsFromAccountToModule(
+			ctx,
+			from, // sender borrowing
+			types.ModuleName,
+			sdk.NewCoins(borrowCoin),
+		); err != nil {
+			return k.updateInsufficientFundsError(ctx, from, amt, err)
+		}
+	}
+
+	// Case #3: NO sender borrow, recipient carry.
+	// Recipient's fractional balance carries over to integer balance by 1.
+	// Always send carry from reserve before receiving borrow from sender to
+	// ensure reserve always has sufficient balance starting from 0.
+	if !senderNeedsBorrow && recipientNeedsCarry {
+		carryCoin := sdk.NewCoin(types.IntegerCoinDenom, sdk.NewInt(1))
+		if err := k.bk.SendCoinsFromModuleToAccount(
+			ctx,
+			types.ModuleName,
+			to, // recipient carrying
+			sdk.NewCoins(carryCoin),
+		); err != nil {
+			// Panic instead of returning error, as this will only error
+			// with invalid state or logic. Reserve should always have
+			// sufficient balance to carry fractional coins.
+			panic(fmt.Errorf("failed to carry fractional coins to %s: %w", to, err))
+		}
+	}
+
```

Comment:
@drklee3 can you add a comment here about how Case #4 is a no-op because the fractional balance sent if fully covered by the user's current fractional balance?